src/tests/e2etest/FirewallTests.cs

```diff
@@ -45,5 +115,91 @@ public void FirewallTest_Regex()
                 RegexAssert.IsMatch(_fingerprintPattern, reported.Fingerprint);
             });
         }
+
+        [Test]
+        public void FirewallTest_SetDesiredRules()
+        {
+            var desired = new Rule[]
+            {
+                new Rule
+                {
+                    DesiredState = DesiredState.Present,
+                    Action = Action.Accept,
+                    Direction = Direction.In,
+                    Protocol = Protocol.Tcp,
+                    SourceAddress = "bing.com",
+                    SourcePort = 1234
+                }
+            };
+
+            Firewall reported = GetReported<Firewall>(_componentName, (Firewall firewall) => (firewall.State != State.Unknown) && (firewall.Fingerprint != null));
+            var initialFingerprint = reported.Fingerprint;
+
+            SetDesired<Rule[]>(_componentName, _desiredRules, desired);
+
+            reported = GetReported<Firewall>(_componentName, (Firewall firewall) => (firewall.Fingerprint != initialFingerprint));
+
+            // Remove the rule added by the test
+            desired = new Rule[]
+            {
+                new Rule
+                {
+                    DesiredState = DesiredState.Absent,
+                    Action = Action.Accept,
+                    Direction = Direction.In,
+                    Protocol = Protocol.Tcp,
+                    SourceAddress = "bing.com",
+                    SourcePort = 1234
+                }
+            };
+            SetDesired<Rule[]>(_componentName, _desiredRules, desired);
+            GetReported<Firewall>(_componentName, (Firewall firewall) => (firewall.Fingerprint != initialFingerprint));
+
+            // Reset the desired state to empty
+            SetDesired<Rule[]>(_componentName, _desiredRules, new Rule[0]);
+
+            Assert.Multiple(() =>
+            {
+                Assert.AreEqual(State.Enabled, reported.State);
+                Assert.AreNotEqual(initialFingerprint, reported.Fingerprint);
```

Comment:
Would it make sense to also have assertions on the rules, in addition to the fingerprint?

Reply:
Yes, however rules are not reported by the Firewall module only the fingerprint is reported as a representation of those rules.